compiler/ilgen/TypeDictionary.cpp

```diff
@@ -285,6 +286,117 @@ StructType::getFieldSymRef(const char *fieldName)
    return (TR::IlReference *)symRef;
    }
 
+class UnionType : public TR::IlType
+   {
+public:
+   TR_ALLOC(TR_Memory::IlGenerator)
+
+   UnionType(const char *name) :
+      TR::IlType(name),
+      _firstField(0),
+      _lastField(0),
+      _size(0),
+      _closed(false),
+      _symRefBV(8 /*initBits*/, TR::comp()->trMemory(), heapAlloc, growable, TR_Memory::BitVector)
+      { }
+
+   TR::DataType getPrimitiveType()                 { return TR::Address; }
+   void Close();
+
+   void AddField(const char *name, TR::IlType *fieldType);
+   TR::IlType * getFieldType(const char *fieldName);
+
+   TR::SymbolReference *getFieldSymRef(const char *name);
+   bool isUnion() { return true; }
+   virtual size_t getSize() { return _size; }
+
+protected:
+   FieldInfo * findField(const char *fieldName);
+
+   FieldInfo * _firstField;
+   FieldInfo * _lastField;
+   size_t      _size;
+   bool        _closed;
+   TR_BitVector _symRefBV;
+   };
+
+void
+UnionType::AddField(const char *name, TR::IlType *typeInfo)
+   {
+   if (_closed)
+      return;
+
+   TR::DataType primitiveType = typeInfo->getPrimitiveType();
+   auto fieldSize = typeInfo->getSize();
+   _size = fieldSize > _size ? fieldSize : _size;
+
+
+   FieldInfo *fieldInfo = new (PERSISTENT_NEW) FieldInfo(name, 0 /* no offset */, typeInfo);
+   if (0 != _lastField)
```

Comment:
for background: JitBuilder currently avoids STL because JitBuilder builds into a static library and it's a lot harder to link against that static library on various platforms if it depends on STL.